pkg/cli/cli.go

```diff
@@ -0,0 +1,263 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cli
+
+import (
+	"errors"
+	"fmt"
+	"log"
+
+	"github.com/spf13/cobra"
+
+	internalconfig "sigs.k8s.io/kubebuilder/internal/config"
+	"sigs.k8s.io/kubebuilder/pkg/cli/internal"
+	"sigs.k8s.io/kubebuilder/pkg/model/config"
+	"sigs.k8s.io/kubebuilder/pkg/plugin"
+)
+
+const (
+	noticeColor = "\033[1;36m%s\033[0m"
+)
+
+// CLI interacts with a command line interface.
+type CLI interface {
+	// Run runs the CLI, usually returning an error if command line configuration
+	// is incorrect.
+	Run() error
+}
+
+// Option is a function that can configure the cli
+type Option func(*cli) error
+
+// cli defines the command line structure and interfaces that are used to
+// scaffold kubebuilder project files.
+type cli struct {
+	// Base command name. Can be injected downstream.
+	commandName string
+	// Default project version. Used in CLI flag setup.
+	defaultProjectVersion string
```

Comment:
Where is this field being used? I think that `DefaultVersion` from `internal/config` may be enough.

Reply:
```go
// WithDefaultProjectVersion is an Option that sets the cli's default project
// version. Setting an unknown version will result in an error.
func WithDefaultProjectVersion(version string) Option {
	return func(c *cli) error {
		c.defaultProjectVersion = version
		return nil
	}
}
```
A downstream project may want to default to another project version.

Reply:
That is setting it, but is it used anywhere? The field is unexported

Reply:
It allows projects using custom `init` logic to inject a default project version so they don't have to set `init --project-version $MY_DEFAULT` every time.

Reply:
My question is, where is that field used. You can set it with the exported function you mention, but then it is not used anywhere, is it? And being an unexported field it can't be used from outside the cli library.

Reply:
It is used to define the `project-version` default in `newInitCmd`:
```go
	cmd.Flags().String("project-version", c.defaultProjectVersion,
		fmt.Sprintf("project version, possible values: (%s)", strings.Join(c.getAvailableProjectVersions(), ", ")))
```

Reply:
That was the part I was missing, mark this as resolved